File: Programmers/Level2/전력망을둘로나누기/전력망을둘로나누기.py
from collections import deque

def bfs(start, visited, graph):
    queue = deque([start])
    visited[start] = True
    result = 1

    while queue:
        v = queue.popleft()
        for i in graph[v]:
            if not visited[i]:
                result += 1
                visited[i] = True
                queue.append(i)
    return result

def solution(n, wires):
    answer = n
    graph = [[] for _ in range(n + 1)]

    for v1, v2 in wires:
        graph[v1].append(v2)
        graph[v2].append(v1)

    for v1, v2 in wires:
        visited = [False] * (n + 1)
        visited[v2] = True
        result = bfs(v1, visited, graph)
        diff = abs(result - (n - result))
        if  diff < answer:
            answer = diff

    return answer

File: Programmers/Level2/전력망을둘로나누기/전력망을둘로나누기.js
function bfs(start, visited, graph) {
    queue = [start];
    visited[start] = true;
    result = 1;

    while (queue.length) {
        let v = queue.pop(0);
        for (let i of graph[v]) {
            if (!visited[i]) {
                result += 1;
                visited[i] = true;
                queue.push(i);
            }
        }
    }
    return result;
}

function solution(n, wires) {
    let answer = n;
    const graph = [...Array(n + 1)].map(() => []);

    for (let [v1, v2] of wires) {
        graph[v1] = [...graph[v1], v2];
        graph[v2] = [...graph[v2], v1];
    }

    for (let [v1, v2] of wires) {
        const visited = Array(n + 1).fill(false);
        visited[v2] = true;
        const result = bfs(v1, visited, graph);
        const diff = Math.abs(result - (n - result));
        if (diff < answer) answer = diff;
    }
    return answer;
}
